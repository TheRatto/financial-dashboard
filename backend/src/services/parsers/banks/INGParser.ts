import { TransactionType } from '@prisma/client';
import { parse } from 'date-fns';

interface BasicParsedTransaction {
  date: Date;
  description: string;
  amount: number;
  type: TransactionType;
  balance: number;
}

export class INGParser {
  name = 'ING';

  canParse(text: string): boolean {
    console.log('\n=== ING Parser Detection ===');
    console.log('Text length:', text.length);
    console.log('First 200 characters:', text.substring(0, 200));
    
    const isING = text.includes('ING');
    console.log('Contains "ING":', isING);
    
    return isING;
  }

  private cleanDescription(rawDesc: string): string {
    const descMatch = rawDesc.match(/((?:BPAY Bill Payment|Salary Deposit|Osko (?:Payment|Deposit)|Internal Transfer|Pay Anyone|EFTPOS Purchase|Visa Purchase|Direct Debit|Deposit|Intl (?:Atmpurchase|Transaction Fee)).*?)(?:\s*-\s*Receipt\s*\d+)?$/);
    return descMatch ? descMatch[1].trim() : '';
  }

  private parseTransactionLine(line: string): BasicParsedTransaction | null {
    const dateRegex = /(\d{2}\/\d{2}\/\d{4})([-]?\d+[,.]?\d+)?([,.]?\d+)?(.*)/;
    const match = line.match(dateRegex);
    
    if (match) {
      const date = parse(match[1], 'dd/MM/yyyy', new Date());
      const description = this.cleanDescription(match[4]);
      const numbers = line.match(/-?\d+[,.]\d+/g);
      if (numbers && numbers.length >= 2) {
        const amount = Math.abs(parseFloat(numbers[0].replace(',', '')));
        const balance = parseFloat(numbers[numbers.length - 1].replace(',', ''));
        const type = line.startsWith('-') ? TransactionType.DEBIT : TransactionType.CREDIT;

        return {
          date,
          description,
          amount,
          type,
          balance
        };
      }
    }
    return null;
  }

  async parse(text: string): Promise<BasicParsedTransaction[]> {
    const transactions: BasicParsedTransaction[] = [];
    const lines = text.split('\n');

    console.log('\n=== ING Parser Starting ===');
    let lineCount = 0;
    let dateMatches = 0;
    
    for (const line of lines) {
      lineCount++;
      
      if (!line.trim() || line.includes('Date') || line.includes('Description')) {
        continue;
      }

      if (line.match(/^\d{2}\/\d{2}\/\d{4}/)) {
        dateMatches++;
        console.log(`\n[Line ${lineCount}] Processing: "${line}"`);
        console.log('Found date:', line.substring(0, 10));
        
        const transaction = this.parseTransactionLine(line);
        if (transaction) {
          transactions.push(transaction);
          console.log('\nParsed Transaction:');
          console.log('------------------');
          console.log(`Date: ${transaction.date.toISOString()}`);
          console.log(`Description: ${transaction.description}`);
          console.log(`Amount: ${transaction.amount} (${transaction.type})`);
          console.log(`Balance: ${transaction.balance}`);
          console.log('------------------');
        }
      }
    }

    console.log('\n=== ING Parser Summary ===');
    console.log('Total lines processed:', lineCount);
    console.log('Lines with date matches:', dateMatches);
    console.log('Transactions successfully parsed:', transactions.length);

    if (transactions.length > 0) {
      console.log('\nFirst transaction:', transactions[0]);
      console.log('Last transaction:', transactions[transactions.length - 1]);
    }

    return transactions;
  }
} 