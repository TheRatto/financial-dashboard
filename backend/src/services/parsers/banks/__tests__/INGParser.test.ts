import { INGParser } from '../INGParser';
import { readFileSync } from 'fs';
import { join } from 'path';
import { TransactionType } from '@prisma/client';

describe('INGParser', () => {
  const parser = new INGParser();

  describe('canParse', () => {
    it('should detect ING credit card statement', async () => {
      const content = readFileSync(
        join(__dirname, '../__fixtures__/ing-credit-sample.txt'),
        'utf8'
      );
      const result = await parser.canParse(content);
      expect(result).toBe(true);
    });

    it('should detect ING savings statement', async () => {
      const content = readFileSync(
        join(__dirname, '../__fixtures__/ing-savings-sample.txt'),
        'utf8'
      );
      const result = await parser.canParse(content);
      expect(result).toBe(true);
    });

    it('should reject non-ING statement', async () => {
      const result = await parser.canParse('Some random text with dates 01/01/2024');
      expect(result).toBe(false);
    });
  });

  describe('parse', () => {
    it('should parse credit card transactions correctly', async () => {
      const content = readFileSync(
        join(__dirname, '../__fixtures__/ing-credit-sample.txt'),
        'utf8'
      );
      const transactions = await parser.parse(content);
      
      expect(transactions).toHaveLength(3);
      expect(transactions[0]).toEqual({
        date: expect.any(Date),
        description: expect.any(String),
        amount: 500,
        type: TransactionType.DEBIT,
        balance: 1500
      });
    });

    it('should parse savings transactions correctly', async () => {
      const content = readFileSync(
        join(__dirname, '../__fixtures__/ing-savings-sample.txt'),
        'utf8'
      );
      const transactions = await parser.parse(content);
      
      expect(transactions).toHaveLength(3);
      expect(transactions[0]).toEqual({
        date: expect.any(Date),
        description: expect.any(String),
        amount: 2500,
        type: TransactionType.CREDIT,
        balance: 15500
      });
    });
  });
}); 