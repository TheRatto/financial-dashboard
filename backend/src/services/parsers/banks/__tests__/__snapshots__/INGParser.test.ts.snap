// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`INGParser should maintain consistent parsing behavior 1`] = `[]`;
